### PR TITLE
fix(#3601): FocusScope wait a frame before reassigning nodeToRestore

### LINKED ISSUE #3601: Dialog focus not restoring to a menu trigger based on react-aria example
# 🐛 Bug Report 

Discussion started here https://github.com/adobe/react-spectrum/pull/3323#issuecomment-1257231881

A Dialog inside a Menu is not restoring focus to the Menu's trigger when the Dialog closes.

## 🔦 Context

Don't know exactly why yet, but in the example, FocusScope's removeTreeNode for the Menu runs before the useLayoutEffect which assigns a nodeToRestore in useRestoreFocus.
https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/focus/src/FocusScope.tsx#L789 is undefined, and I don't think it'd matter if we checked if this was a restoring scope instead of nodeToRestore existing, it'd be overwritten in the useRestoreFocus layoutEffect to the wrong value.

In our component, it's reversed. There seems to be a substantial delay between closing of the menu and actually unmounting it. However, I don't quite know where that delay is coming from, except that we use Transition instead of CSSTransition. I don't know if that is what makes the difference.

I pushed a branch with the sandbox code added to one of our stories for easier debugging https://github.com/adobe/react-spectrum/compare/focusscope-nodetorestore-wrong?expand=1

Reporter:
> I tried the new release and the issue seems to persist, I updated the sandbox. Also I removed CSSTransition to rule it out. I wonder if I constructed the test case wrong, I copied the files from the two official styled examples but some imports were changed so there may potentially be some API changes that didn't make it to the sandbox. The reason I doubt it is because the release announcement example on Twitter seems to be working nicely.

## 💻 Code Sample

https://codesandbox.io/s/sad-platform-58uy0k?file=/src/App.tsx

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   |
| Browser          |
| Operating System |

### PR DESCRIPTION
**Pull request by majornista**
So that a `FocusScope` opening from a `FocusScope` that is closing does not overwrite the `nodeToRestore` we have reassigned to it with a `nodeToRestore` within the `FocusScope` that is closing, we need to wait a frame.

Closes #3601

## ✅ Pull Request Checklist:

- [x] Included link to corresponding React Spectrum GitHub Issue #3601.
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

This updated Tailwind CSS example demonstrates the corrected behavior: https://codesandbox.io/s/exciting-hawking-dsnlzr.

Storybook example: https://reactspectrum.blob.core.windows.net/reactspectrum/792a34d84d14815d733c9a15579c944c30b2cfd6/storybook/index.html?path=/story/focusscope--restore-focus&providerSwitcher-toastPosition=bottom

## 🧢 Your Project:

Adobe/Accessibility

### CODE REVIEW

**Comment by frontsideair** [comment]
It seems to be working even with `CSSTransitions`s removed, I would like to try in my app when there's a prerelease.

**Review by ktabors — COMMENTED** [review]
Could we add a FocusScope story with this example?

**Comment by majornista** [comment]
> Could we add a FocusScope story with this example?

Added a storybook example: https://reactspectrum.blob.core.windows.net/reactspectrum/cb8eaed90d6baa3e0c1f407bfb51ab2f3b3e1144/storybook/index.html?path=/story/focusscope--restore-focus&providerSwitcher-toastPosition=bottom

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 1]
```diff
@@ -788,11 +788,18 @@ class Tree {
...
         current !== node &&
         node.nodeToRestore &&
-        current.nodeToRestore &&
-        node.scopeRef.current &&
-        isElementInScope(current.nodeToRestore, node.scopeRef.current)
+        (
+          // current.nodeToRestore might be undefined when the current node has just been added.
+          !current.nodeToRestore ||
+          (
+            current.nodeToRestore &&
```
Comment (carolframpton) [inline packages/@react-aria/focus/src/FocusScope.tsx:795]: Do you have to check for `current.nodeToRestore`? Isn't it implied on this side of the `||`?

**Comment by carolframpton** [comment]
@majornista I don't feel qualified to review any focus code. We're having focus issues so I was just looking around.

**Comment by majornista** [comment]
> @majornista I don't feel qualified to review any focus code. We're having focus issues so I was just looking around.

No worries. I didn't mean to put you on the spot. By refreshing your review comment, I just to let you know that I appreciated your feedback and have incorporated your suggested change.

**Comment by snowystinger** [comment]
This got fixed somewhere along the way. I can no longer reproduce the problem with our latest code

**Pull request closed without merging**